### PR TITLE
[IMP] product: remove detailed_type field

### PR DESCRIPTION
**Pull request by kcv-odoo**
### [IMP] {website_,}event{,booth}_sale: remove product types
Purpose of this commit to clean up product types.

In this commit remove `event` and `event_booth` type and
use service type and move those types in to
`service_tracking` field so now service type field with
`service_tracking` equals to event would be event type
product.

### [IMP] product: remove detailed_type field
Purpose of this commit to remove `detailed_type` field.

Since remove of different kind of types from detailed_type
field `detailed_type` field will contains same type as
`type` field so we can remove `detailed_type` field and
use type field as radio button in UX.

task-3938213

### CODE REVIEW

File: addons/event_booth_sale/models/product.py [hunk 1]
```diff
@@ -7,25 +7,20 @@
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
-    detailed_type = fields.Selection(selection_add=[
+    service_tracking = fields.Selection(selection_add=[
         ('event_booth', 'Event Booth'),
-    ], ondelete={'event_booth': 'set service'})
+    ], ondelete={'event_booth': 'set no'})
```
Comment (Feyensv) [inline addons/event_booth_sale/models/product.py:12]:
```suggestion
    ], ondelete={'event_booth': 'set default'})
```
There is a generic word to specify the fallback on default value ;).

File: addons/sale_project/models/product_template.py [hunk 2]
```diff
@@ -19,17 +19,12 @@ def _selection_service_policy(self):
         return service_policies
 
     service_tracking = fields.Selection(
-        selection=[
-            ('no', 'Nothing'),
+        selection_add=[
```
Comment (Feyensv) [inline addons/sale_project/models/product_template.py:22]: missing ondelete specification

**Comment by kcv-odoo** [comment]
Drop removing `type` or `detailed_type` field for now will do in other PR

File: addons/repair/views/product_views.xml [hunk 3]
```diff
@@ -7,7 +7,7 @@
         <field name="inherit_id" ref="product.product_template_form_view"/>
         <field name="arch" type="xml">
             <group name="group_general" position="inside">
-                <field name="create_repair" invisible="detailed_type not in ('consu', 'service')"/>
+                <field name="create_repair" invisible="not is_storable"/>
```
Comment (Feyensv) [inline addons/repair/views/product_views.xml:10]: doesn't seem to match the same conditions as before :eyes:

File: addons/sale/models/product_template.py [hunk 4]
```diff
@@ -19,6 +19,22 @@ class ProductTemplate(models.Model):
...
+        default="no",
+        compute="_compute_service_tracking",
+        store=True,
+        readonly=False,
+        help="Link this product to different type of service. \
+        For project/task related Types: \
+        On Sales order confirmation, this product can generate a project and/or task. \
+        From those, you can track the service you are selling.\n \
+        'In sale order\'s project': Will use the sale order\'s configured project if defined or fallback to \
+        creating a new project based on the selected template.\
```
Comment (Feyensv) [inline addons/sale/models/product_template.py:36]: This help must be adapted.
Who decided the string of the field? We won't use `Create on Order` anymore ?

File: addons/sale/views/product_views.xml [hunk 5]
```diff
@@ -13,8 +13,9 @@
                 <field name="service_type" widget="radio" invisible="True"/>
                 <field name="visible_expense_policy" invisible="1"/>
             </field>
-            <field name="detailed_type" position="after">
+            <field name="type" position="after">
                 <field name="invoice_policy" required="1"/>
+                <field name="service_tracking" required="1" invisible="1 or type != 'service' or not sale_ok"/>
```
Comment (Feyensv) [inline addons/sale/views/product_views.xml:18]: make it required on the model directly, and handle it in the upgrade script.

File: addons/sale_project/models/sale_order.py [hunk 6]
```diff
@@ -46,12 +46,15 @@ def _compute_show_project_and_task_button(self):
...
+            ('product_id.type', '=', 'service'),
         ], aggregates=['order_id:array_agg'])[0][0]
         for order in self:
             order.show_project_button = order.id in show_button_ids and order.project_count
             order.show_task_button = order.show_project_button or order.tasks_count
-            order.show_create_project_button = is_project_manager and order.id in show_button_ids and not order.project_count and 'service' in order.order_line.product_template_id.mapped('detailed_type')
+            order.show_create_project_button = is_project_manager\
+                and order.id in show_button_ids\
+                and not order.project_count\
+                and 'service' in order.order_line.mapped('product_type')
```
Comment (Feyensv) [inline addons/sale_project/models/sale_order.py:57]:
```suggestion
            order.show_create_project_button = (
                is_project_manager
                and order.id in show_button_ids
                and not order.project_count
            )
```
AFAICS, the product type is already checked by `order.id in show_button_ids` :thinking:

File: addons/sale_project/models/sale_order.py [hunk 7]
```diff
@@ -225,7 +228,10 @@ def action_create_project(self):
             }
 
         sorted_line = self.order_line.sorted('sequence')
-        default_sale_line = next((sol for sol in sorted_line if sol.product_id.detailed_type == 'service' and not sol.is_downpayment), self.env['sale.order.line'])
+        default_sale_line = next((
+            sol for sol in sorted_line if sol.product_id.type == 'service'
+                and not sol.is_downpayment
+        ), self.env['sale.order.line'])
```
Comment (Feyensv) [inline addons/sale_project/models/sale_order.py:234]:
```suggestion
        default_sale_line = next((
            sol for sol in sorted_line
            if sol.product_id.type == 'service' and not sol.is_downpayment
        ), self.env['sale.order.line'])
```

File: addons/sale_timesheet/__init__.py [hunk 8]
```diff
@@ -10,7 +10,11 @@ def uninstall_hook(env):
     env.ref("account.account_analytic_line_rule_billing_user").write({'domain_force': "[(1, '=', 1)]"})
 
 def _sale_timesheet_post_init(env):
-    products = env['product.template'].search([('detailed_type', '=', 'service'), ('invoice_policy', '=', 'order'), ('service_type', '=', 'manual')])
+    products = env['product.template'].search([
+        ('type', '=', 'service'),
+        ('invoice_policy', '=', 'order'),
+        ('service_type', '=', 'manual'),
+    ])
```
Comment (Feyensv) [inline addons/sale_timesheet/__init__.py:17]: Mhm, seems wrong, event booth, tickets, booking fees, ... will be set as `timesheet` here :thinking:

File: addons/stock/models/stock_move.py [hunk 9]
```diff
@@ -209,7 +209,8 @@ def _compute_location_dest_id(self):
                 location_dest = move.location_final_id
             move.location_dest_id = location_dest
 
-    @api.depends('has_tracking', 'picking_type_id.use_create_lots', 'picking_type_id.use_existing_lots')
+    @api.depends(
+            'has_tracking', 'picking_type_id.use_create_lots', 'picking_type_id.use_existing_lots')
```
Comment (Feyensv) [inline addons/stock/models/stock_move.py:213]: unrelated diff

File: addons/website_event_booth_sale/tests/test_event_booth_sale.py [hunk 10]
```diff
@@ -85,3 +91,39 @@ def test_booth_pricelists_different_currencies(self):
...
+        })
+        product_service = self.env['product.product'].create({
+            'name': 'Can be zero price',
+            'type': 'service',
+            'service_tracking': 'event_booth',
+            'list_price': 0,
+            'website_published': True,
+        })
+
+        with patch.object(ProductTemplate, '_get_product_types_allow_zero_price', lambda pt: ['event_booth']):
```
Comment (Feyensv) [inline addons/website_event_booth_sale/tests/test_event_booth_sale.py:122]: If you move the test here, there is no need for this patch anymore.
You should keep it `website_sale`, make it so that the patch return 'no', so that any zero priced product can be added, and check that without the patch, they cannot.

File: addons/website_sale/models/product_template.py [hunk 11]
```diff
@@ -692,7 +692,7 @@ def _get_alternative_product_filter(self):
     @api.model
     def _get_product_types_allow_zero_price(self):
         """
-        Returns a list of detailed types (`product.template.detailed_type`) that can ignore the
+        Returns a list of service_tracing (`product.template.service_tracking`) that can ignore the
```
Comment (Feyensv) [inline addons/website_sale/models/product_template.py:695]:
```suggestion
        Returns a list of service_tracking (`product.template.service_tracking`) that can ignore the
```

File: addons/website_sale_slides/views/product_template_views.xml [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="product_template_form_view" model="ir.ui.view">
+        <field name="name">product.template.inherit</field>
```
Comment (Feyensv) [inline addons/website_sale_slides/views/product_template_views.xml:4]:
```suggestion
        <field name="name">product.template.inherit.website.sale.slides</field>
```

File: addons/website_sale_slides/views/slide_channel_views.xml [hunk 13]
```diff
@@ -59,7 +67,15 @@
...
-                context="{'default_detailed_type': 'course', 'default_invoice_policy': 'order', 'default_purchase_ok': False, 'default_sale_ok': True, 'default_website_published': True}"/>
+                context="{
+                    'default_type': 'service',
+                    'default_service_tracking': 'course',
+                    'default_invoice_policy': 'order',
+                    'default_purchase_ok': False,
+                    'default_sale_ok': True,
+                    'default_website_published': True
+                }"
+            />
```
Comment (Feyensv) [inline addons/website_sale_slides/views/slide_channel_views.xml:78]: you should add one level of indentation.

File: addons/sale/models/product_template.py (hunk 4 above)
Comment (kcv-odoo) [reply addons/sale/models/product_template.py:36]: Reverted string!
Waiting for help string from @dalaOdoo

File: addons/website_sale_slides/models/product_template.py [hunk 14]
```diff
@@ -7,14 +7,9 @@
...
-    detailed_type = fields.Selection(selection_add=[
-        ('course', 'Course'),
-    ], ondelete={'course': 'set service'})
-
-    def _detailed_type_mapping(self):
-        type_mapping = super(ProductTemplate, self)._detailed_type_mapping()
-        type_mapping['course'] = 'service'
-        return type_mapping
+    service_tracking = fields.Selection(selection_add=[
+        ('course', 'Course Attendence'),
```
Comment (Feyensv) [inline addons/website_sale_slides/models/product_template.py:11]:
```suggestion
        ('course', 'Course Attendance'),
```

File: addons/sale_timesheet/__init__.py [hunk 15]
```diff
@@ -10,7 +10,12 @@ def uninstall_hook(env):
     env.ref("account.account_analytic_line_rule_billing_user").write({'domain_force': "[(1, '=', 1)]"})
 
 def _sale_timesheet_post_init(env):
-    products = env['product.template'].search([('detailed_type', '=', 'service'), ('invoice_policy', '=', 'order'), ('service_type', '=', 'manual')])
+    products = env['product.template'].search([
+        ('type', '=', 'service'),
+        ('service_tracking', 'in', ['task_global_project', 'task_in_project', 'project_only']),
+        ('invoice_policy', '=', 'order'),
+        ('service_type', '=', 'manual'),
+    ])
```
[the remark this answers is not in the document]
Comment (xavierbol) [reply addons/sale_timesheet/__init__.py:18]: Good question. 🤔 
I think we can indeed include the products whose `service_tracking` is `no`. 🙂

**Review by xavierbol — APPROVED** [review]
Nice work! 💪🏼 

LGTM for the service part, thanks for your work! 🙂

**Review by Feyensv — APPROVED** [review]

**Comment by Feyensv** [comment]
@robodoo r+

**Comment by Demesmaeker** [comment]
@robodoo r+
